applications/crossbar/src/modules/cb_port_requests.erl

```diff
@@ -1602,3 +1498,72 @@ save_phone_numbers_doc(Context, JObj) ->
             lager:error("failed to save phone_numbers doc in ~s : ~p", [AccountId, _Status]),
             'error'
     end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% @end
+%%--------------------------------------------------------------------
+-spec generate_loa_from_port(cb_context:context(), wh_json:object()) ->
+                                    cb_context:context().
+generate_loa_from_port(Context, PortRequest) ->
+    AccountId = cb_context:account_id(Context),
+
+    ResellerId = wh_services:find_reseller_id(AccountId),
+    ResellerDoc = cb_context:account_doc(cb_context:set_account_id(Context, ResellerId)),
+
+    AccountDoc = cb_context:account_doc(Context),
+
+    Numbers = [wnm_util:pretty_print(N) || N <- wh_json:get_keys(<<"numbers">>, PortRequest)],
+
+    QRCode = create_qr_code(cb_context:account_id(Context), wh_doc:id(PortRequest)),
+
+    generate_loa_from_template(Context
+                               ,props:filter_undefined(
+                                  [{<<"reseller">>, wh_json:to_proplist(ResellerDoc)}
+                                   ,{<<"account">>, wh_json:to_proplist(AccountDoc)}
+                                   ,{<<"numbers">>, Numbers}
+                                   ,{<<"bill">>, wh_json:to_proplist(wh_json:get_value(<<"bill">>, PortRequest, wh_json:new()))}
+                                   ,{<<"request">>, wh_json:to_proplist(PortRequest)}
+                                   ,{<<"qr_code">>, QRCode}
+                                   ,{<<"type">>, <<"loa">>}
+                                  ])
+                               ,ResellerId
+                               ,wh_json:get_value(<<"carrier">>, PortRequest)
+                              ).
+
+-spec generate_loa_from_template(cb_context:context(), wh_proplist(), ne_binary(), api_binary()) ->
+                                        cb_context:context().
+generate_loa_from_template(Context, TemplateData, ResellerId, Carrier) ->
+    Template = find_template(ResellerId, Carrier),
+    case kz_pdf:generate(ResellerId, TemplateData, Template) of
+        {'error', _R} -> cb_context:set_resp_status(Context, 'error');
+        {'ok', PDF} ->
+            cb_context:set_resp_status(
+              cb_context:set_resp_data(Context, PDF)
+              ,'success'
+             )
+    end.
+
+-spec create_qr_code(api_binary(), api_binary()) -> wh_proplist() | 'undefined'.
+create_qr_code('undefined', _) -> 'undefined';
+create_qr_code(_, 'undefined') -> 'undefined';
+create_qr_code(AccountId, PortRequestId) ->
+    lager:debug("create qr code for ~s - ~s", [AccountId, PortRequestId]),
+    CHL = <<AccountId/binary, "-", PortRequestId/binary>>,
+    Url = <<"https://chart.googleapis.com/chart?chs=300x300&cht=qr&chl=", CHL/binary, "&choe=UTF-8">>,
+
+    case ibrowse:send_req(wh_util:to_list(Url)
```

Comment:
@k-anderson can we move this to a common place ? maybe wh_util ? or kz_image (manipulate and generate images) ? there is also a create_qr_code in notify whapp.

also, maybe add a type (port_request or other name in this case) to help differentiate origins ?